From e313c6a5e7c6355984f38d400e718ce1b506e6da Mon Sep 17 00:00:00 2001
From: daniel wappner <dwappner@dc.uba.ar>
Date: Fri, 31 Jul 2026 10:52:04 -0300
Subject: [PATCH] reintroduce handling of non-pointers in
 getOneLevelOffsetsFromAllValues

---
 .../Instrumentation/SanitizerCoverage.cpp     | 23 ++++++++++---------
 1 file changed, 12 insertions(+), 11 deletions(-)

diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index 499ed8f54..11d7d86d6 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -1179,17 +1179,18 @@ getOneLevelOffsetsFromAllValues(ArrayRef<Value *> PossibleStoredValues) {
       // Depth 1: struct possibly containing function pointers
       return getOffsetsFromShallowStructType(STy);
     }
-    if (StoredType->isPointerTy()) {
-      Type *PointeeType = StoredType->getPointerElementType();
-      if (PointeeType->isFunctionTy()) {
-        // Depth 0: naked function pointer
-        return getOffsetsFromFunctionType(PointeeType);
-      } // else: must be struct
-      if (StructType *STy = dyn_cast<StructType>(PointeeType)) {
-        // Depth 1 again: pointer to struct possibly containing function
-        // pointers
-        return getOffsetsFromPointerToShallowStructType(STy);
-      }
+    if (!StoredType->isPointerTy())
+      continue;
+    // else: assume Pointer
+    Type *PointeeType = StoredType->getPointerElementType();
+    if (PointeeType->isFunctionTy()) {
+      // Depth 0: naked function pointer
+      return getOffsetsFromFunctionType(PointeeType);
+    } // else: must be struct
+    if (StructType *STy = dyn_cast<StructType>(PointeeType)) {
+      // Depth 1 again: pointer to struct possibly containing function
+      // pointers
+      return getOffsetsFromPointerToShallowStructType(STy);
     }
     // TODO: handle more types
     assert(false && "Unexpected type during function pointer analysis.\n");
-- 
2.43.0

