From bc250443cee510279caf131bd4620c444e8b38fa Mon Sep 17 00:00:00 2001
From: daniel wappner <dwappner@dc.uba.ar>
Date: Mon, 3 Aug 2026 18:06:19 -0300
Subject: [PATCH 17/18] remove safety check that is too restrictive

---
 llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp | 2 --
 1 file changed, 2 deletions(-)

diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index efad69b4393b..c2e8907428d4 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -1268,9 +1268,7 @@ void ModuleSanitizerCoverage::InjectFuncPointerCallbackUsingOffsetChain(
      */
     assert(StoredValue->getType()->isPointerTy());
     Type *DesiredFType = TypeChain[0];
-    Type *MiscPointeeType = StoredValue->getType()->getPointerElementType();
     assert(DesiredFType->isFunctionTy());
-    assert(MiscPointeeType == DesiredFType);
     InjectFuncPointerCallback(StorePtr, StoredValue, Builder);
   } else {
     /*
-- 
2.43.0

