From 611f586d573c4f7f14de2f8a52ac6a318a80b2b9 Mon Sep 17 00:00:00 2001
From: Daniel Wappner <dwappner@dc.uba.ar>
Date: Tue, 4 Aug 2026 15:53:34 -0300
Subject: [PATCH 18/18] inject NULL and ERR_PTR check before struct pointer
 dereference

---
 .../Instrumentation/SanitizerCoverage.cpp     | 121 ++++++++++--------
 1 file changed, 67 insertions(+), 54 deletions(-)

diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index c2e8907428d4..33b2b86ff508 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -11,6 +11,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "llvm/Transforms/Instrumentation/SanitizerCoverage.h"
+#include "llvm/ADT/APInt.h"
 #include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringRef.h"
@@ -19,6 +20,7 @@
 #include "llvm/IR/BasicBlock.h"
 #include "llvm/IR/CFG.h"
 #include "llvm/IR/Constant.h"
+#include "llvm/IR/Constants.h"
 #include "llvm/IR/DataLayout.h"
 #include "llvm/IR/DebugInfo.h"
 #include "llvm/IR/DerivedTypes.h"
@@ -28,6 +30,7 @@
 #include "llvm/IR/IRBuilder.h"
 #include "llvm/IR/InlineAsm.h"
 #include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
 #include "llvm/IR/IntrinsicInst.h"
 #include "llvm/IR/Intrinsics.h"
 #include "llvm/IR/LLVMContext.h"
@@ -287,18 +290,10 @@ private:
       ArrayRef<TraceInstrumentTarget> BlocksToInstrument,
       json::Object &DebugDump, json::Array &StoreInstInfoArray,
       unsigned InstrumentedCount, bool HaveFuncPointerArgTypes);
-  void InjectFuncPointerCallbackUsingOffsetChain(Value *StorePtr,
-                                                 Value *StoredValue,
-                                                 Offsets OffsetChain,
-                                                 Types TypeChain,
-                                                 IRBuilder<> &Builder);
-  void InjectFuncPointerCallback(Value *StorePtr, Value *StoredValue,
-                                 IRBuilder<> &Builder);
-  void InjectFuncPointerCallbackAtShallowStruct(Value *StorePtr,
-                                                Value *StoredValue,
-                                                Offsets OffsetChain,
-                                                Types TypeChain,
-                                                IRBuilder<> &Builder);
+  void InjectFuncPointerCallbackUsingOffsetChain(TraceInstrumentTarget Target);
+  void InjectFuncPointerCallbackAtShallowStruct(TraceInstrumentTarget Target);
+  CallInst *InjectFuncPointerCallback(Value *StorePtr, Value *StoredValue,
+                                      IRBuilder<> &Builder);
   void InjectTraceForSwitch(Function &F,
                             ArrayRef<Instruction *> SwitchTraceTargets);
   bool InjectCoverage(Function &F, ArrayRef<BasicBlock *> AllBlocks,
@@ -1186,7 +1181,7 @@ getOneLevelOffsetsFromAllValues(ArrayRef<Value *> PossibleStoredValues) {
     // else: assume Pointer
     Type *PointeeType = StoredType->getPointerElementType();
     if (PointeeType->isIntegerTy())
-      //pointer to int is usually not a type signature we can get function pointers from
+      // pointer to int is usually not a type signature we can get function pointers from
       continue;
     if (PointeeType->isFunctionTy()) {
       // Depth 0: naked function pointer
@@ -1257,19 +1252,24 @@ void fillJsonWithFunctionInfo(Function &F, json::Object &DebugDump) {
 } // namespace
 
 void ModuleSanitizerCoverage::InjectFuncPointerCallbackUsingOffsetChain(
-    Value *StorePtr, Value *StoredValue, Offsets OffsetChain, Types TypeChain,
-    IRBuilder<> &Builder) {
+    TraceInstrumentTarget Target) {
 
-  if (OffsetChain.size() == 1) {
+  if (Target.OffsetChain.size() == 1) {
     /*
      * Depth 0: Naked function pointer.
      * We return the original StorePtr as the address and the function pointer
      * as the value.
      */
-    assert(StoredValue->getType()->isPointerTy());
-    Type *DesiredFType = TypeChain[0];
+    assert(Target.SI->getValueOperand()->getType()->isPointerTy());
+    Type *DesiredFType = Target.TypeChain[0];
     assert(DesiredFType->isFunctionTy());
-    InjectFuncPointerCallback(StorePtr, StoredValue, Builder);
+
+    IRBuilder<> IRB(Target.BB->getTerminator());
+    if (Target.SI->getDebugLoc())
+      IRB.SetCurrentDebugLocation(Target.SI->getDebugLoc());
+
+    InjectFuncPointerCallback(Target.SI->getPointerOperand(),
+                              Target.SI->getValueOperand(), IRB);
   } else {
     /*
      * Depth 1: Struct containing function pointers.
@@ -1277,10 +1277,9 @@ void ModuleSanitizerCoverage::InjectFuncPointerCallbackUsingOffsetChain(
      * struct, and load it to get the actual function pointer.
      */
     assert(
-        OffsetChain.size() == 2 &&
+        Target.OffsetChain.size() == 2 &&
         "Currently only Offset chain of size 2 (depth 1 structs) is supported");
-    InjectFuncPointerCallbackAtShallowStruct(StorePtr, StoredValue, OffsetChain,
-                                             TypeChain, Builder);
+    InjectFuncPointerCallbackAtShallowStruct(Target);
   }
 }
 
@@ -1288,24 +1287,33 @@ void ModuleSanitizerCoverage::InjectFuncPointerCallbackUsingOffsetChain(
 // parameters:
 //   - address of the variable being written to,
 //   - function pointer reachable through the value written.
-void ModuleSanitizerCoverage::InjectFuncPointerCallback(Value *StorePtr,
-                                                        Value *StoredValue,
-                                                        IRBuilder<> &Builder) {
-  Builder.CreateCall(SanCovStoreFPointerFunction,
-                     {
-                         Builder.CreatePointerCast(StorePtr, IntptrTy),
-                         Builder.CreatePointerCast(StoredValue, IntptrTy),
-                     });
+CallInst *ModuleSanitizerCoverage::InjectFuncPointerCallback(
+    Value *StorePtr, Value *StoredValue, IRBuilder<> &Builder) {
+  CallInst *Res =
+      Builder.CreateCall(SanCovStoreFPointerFunction,
+                         {
+                             Builder.CreatePointerCast(StorePtr, IntptrTy),
+                             Builder.CreatePointerCast(StoredValue, IntptrTy),
+                         });
+  Res->setCannotMerge(); // gets the PC using GET_CALLER_PC.
+  return Res;
 }
 
 void ModuleSanitizerCoverage::InjectFuncPointerCallbackAtShallowStruct(
-    Value *StorePtr, Value *StoredValue, Offsets OffsetChain, Types TypeChain,
-    IRBuilder<> &Builder) {
+    TraceInstrumentTarget Target) {
+  unsigned long LINUX_MAX_ERRNO = 4095;
 
-  StoredValue = StoredValue->stripPointerCastsAndAliases();
-  unsigned Offset = OffsetChain[0];
-  Type *ContainedType = TypeChain[0];
-  Type *StructType = TypeChain[1];
+  Value *StoredValue = Target.SI->getValueOperand()->stripPointerCastsAndAliases();
+  Value *StorePtr = Target.SI->getPointerOperand();
+  unsigned Offset = Target.OffsetChain[0];
+  Type *ContainedType = Target.TypeChain[0];
+  Type *StructType = Target.TypeChain[1];
+
+  Instruction *Terminator = Target.BB->getTerminator();
+  IRBuilder<> IRB(Terminator);
+  DebugLoc EntryLoc = Target.SI->getDebugLoc();
+  if (EntryLoc)
+    IRB.SetCurrentDebugLocation(EntryLoc);
 
   unsigned AddrSpace = StoredValue->getType()->getPointerAddressSpace();
   PointerType *StructPtrTy = PointerType::get(StructType, AddrSpace);
@@ -1313,30 +1321,35 @@ void ModuleSanitizerCoverage::InjectFuncPointerCallbackAtShallowStruct(
   // This cast should only be necessary because this is the first
   // ever Value in the nested struct chain. The recursive version
   // should only do this cast before traversing the offset chain.
-  Value *Cast = Builder.CreatePointerCast(StoredValue, StructPtrTy);
-
-  Value *ContainedPtrAddress = Builder.CreateInBoundsGEP(
-      StructType, Cast,
+  StoredValue = IRB.CreatePointerCast(StoredValue, StructPtrTy);
+
+  // The hacky hack here is that NULL=0 and ERR_PTR values live in the
+  // [max(unsigned long)-MAX_ERRNO, max(unsigned long)] space.
+  // So if BasePointer is NULL and we let it underflow, we get something in
+  // [max(unsigned long)-MAX_ERRNO-1, max(unsigned long)] for both NULL and ERR_PTR.
+  Value *BasePointer = IRB.CreatePtrToInt(StoredValue, IntptrTy);
+  Value *PossibleErrorPtr = IRB.CreateSub(BasePointer, ConstantInt::get(IntptrTy, 1));
+  Value *IsError = IRB.CreateICmpUGE(
+    PossibleErrorPtr,
+    ConstantInt::get(IntptrTy,
+      APInt::getMaxValue(IntptrTy->getIntegerBitWidth()) - (LINUX_MAX_ERRNO + 1)
+    )
+  );
+  Instruction *ThenInst = SplitBlockAndInsertIfThen(IsError, Target.BB->getTerminator(), false);
+
+  // Populate safeguarded block with the callback
+  IRB.SetInsertPoint(ThenInst);
+  Value *ContainedPtrAddress = IRB.CreateInBoundsGEP(
+      StructType, StoredValue,
       {ConstantInt::get(Type::getInt32Ty(StoredValue->getContext()), 0),
        ConstantInt::get(Type::getInt32Ty(StoredValue->getContext()), Offset)});
-
-  // This is the injected Load that potentially accesses forbidden memory.
-  // We just checked that the Type structure of StoredValue leads to this
-  // address existing. We haven't checked that StoredValue is properly
-  // initialized/allocated according to its type.
-  Value *FunctionPtr = Builder.CreateLoad(ContainedType, ContainedPtrAddress);
-  InjectFuncPointerCallback(StorePtr, FunctionPtr, Builder);
+  Value *FunctionPtr = IRB.CreateLoad(ContainedType, ContainedPtrAddress);
+  InjectFuncPointerCallback(StorePtr, FunctionPtr, IRB);
 }
 
 void ModuleSanitizerCoverage::InjectFuncPointerTraceAtBlock(
     TraceInstrumentTarget Target) {
-  Instruction *Terminator = Target.BB->getTerminator();
-  DebugLoc EntryLoc = Terminator->getDebugLoc();
-  IRBuilder<> IRB(Terminator);
-  IRB.SetCurrentDebugLocation(EntryLoc);
-  InjectFuncPointerCallbackUsingOffsetChain(
-      Target.SI->getPointerOperand(), Target.SI->getValueOperand(),
-      Target.OffsetChain, Target.TypeChain, IRB);
+  InjectFuncPointerCallbackUsingOffsetChain(Target);
 }
 
 void ModuleSanitizerCoverage::InjectFuncPointerTracesAtBlocks(
-- 
2.43.0

