From 73afeb462a46834b413049d4cf9b84329a8798a2 Mon Sep 17 00:00:00 2001
From: Daniel Wappner <dwappner@dc.uba.ar>
Date: Tue, 31 Mar 2026 15:33:17 -0300
Subject: [PATCH 2/3] move things around

---
 .../Instrumentation/SanitizerCoverage.cpp     | 56 +++++++++----------
 1 file changed, 27 insertions(+), 29 deletions(-)

diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index b7327ce7a8b1..b630fed705c9 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -653,31 +653,6 @@ static bool IsInterestingCmp(ICmpInst *CMP, const DominatorTree *DT,
   return true;
 }
 
-/**
-/ Returns true only where the stored value is a pointer to a function.
-/ The stored value can be a statically known function, or it can come
-/ from somewhere like the field of a struct in allocated memory.
-/
-/ Looking at the Type of the stored Value by doing something like
-/  Store->getValueOperand()->stripPointerCasts()->getType()
-/  will fail to identify dynamic functions as a FunctionType because
-/  LLVM's FunctionType refers to statically known global functions,
-/  which the dynamic function by definition cannot be.
-/
-/ This is why we try to use debug information on the stored variable's
-/ supposed type instead.
-/
-/ TO DO: actually attempt the version that "doesn't work"
-**/
-static bool IsInterestingStore(StoreInst *Store) {
-
-  Value *StoreAddr = Store->getPointerOperand();
-
-  if () // To DO: use debug information
-    return true;
-  return false;
-}
-
 void ModuleSanitizerCoverage::instrumentFunction(
     Function &F, DomTreeCallback DTCallback, PostDomTreeCallback PDTCallback) {
   if (F.empty())
@@ -719,7 +694,6 @@ void ModuleSanitizerCoverage::instrumentFunction(
   SmallVector<GetElementPtrInst *, 8> GepTraceTargets;
   SmallVector<LoadInst *, 8> Loads;
   SmallVector<StoreInst *, 8> Stores;
-  SmallVector<StoreInst *, 8> FunctionPointerStores;
 
   const DominatorTree *DT = DTCallback(F);
   const PostDominatorTree *PDT = PDTCallback(F);
@@ -757,8 +731,7 @@ void ModuleSanitizerCoverage::instrumentFunction(
           Stores.push_back(SI);
       if (Options.TraceFunctionPointerStores)
         if (StoreInst *SI = dyn_cast<StoreInst>(&Inst))
-          if (IsInterestingStore(SI))
-            FunctionPointerStores.push_back(SI);
+          Stores.push_back(SI);
       if (Options.StackDepth)
         if (isa<InvokeInst>(Inst) ||
             (isa<CallInst>(Inst) && !isa<IntrinsicInst>(Inst)))
@@ -773,7 +746,7 @@ void ModuleSanitizerCoverage::instrumentFunction(
   InjectTraceForDiv(F, DivTraceTargets);
   InjectTraceForGep(F, GepTraceTargets);
   InjectTraceForLoadsAndStores(F, Loads, Stores);
-  InjectTraceForFunctionPointerStores(F, FunctionPointerStores);
+  InjectTraceForFunctionPointerStores(F, Stores);
 }
 
 GlobalVariable *ModuleSanitizerCoverage::CreateFunctionLocalArrayInSection(
@@ -992,6 +965,31 @@ void ModuleSanitizerCoverage::InjectTraceForLoadsAndStores(
   }
 }
 
+/**
+/ Returns true only where the stored value is a pointer to a function.
+/ The stored value can be a statically known function, or it can come
+/ from somewhere like the field of a struct in allocated memory.
+/
+/ Looking at the Type of the stored Value by doing something like
+/  Store->getValueOperand()->stripPointerCasts()->getType()
+/  will fail to identify dynamic functions as a FunctionType because
+/  LLVM's FunctionType refers to statically known global functions,
+/  which the dynamic function by definition cannot be.
+/
+/ This is why we try to use debug information on the stored variable's
+/ supposed type instead.
+/
+/ TO DO: actually attempt the version that "doesn't work"
+**/
+static bool IsInterestingStore(StoreInst *Store){
+
+  Value *StoreAddr = Store->getPointerOperand();
+
+  if () //To DO: use debug information
+      return true;
+  return false;
+}
+
 // On every store to a function pointer variable call we call a run-time
 // function __sanitizer_cov_store_fun_pointer* with two parameters:
 //   - address of te variable,
-- 
2.43.0

