From 0e93398562f54ca43773ae55c36448edad1c1b3d Mon Sep 17 00:00:00 2001
From: Daniel Wappner <dwappner@dc.uba.ar>
Date: Tue, 31 Mar 2026 22:59:18 -0300
Subject: [PATCH] fix things compiler complains about

---
 .../Instrumentation/SanitizerCoverage.cpp     | 45 +++++--------------
 1 file changed, 10 insertions(+), 35 deletions(-)

diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index e77f0a9eb83e..d281f8d5610a 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -510,7 +510,7 @@ bool ModuleSanitizerCoverage::instrumentModule(
       M.getOrInsertFunction(SanCovStore16, VoidTy, Int128PtrTy);
 
   SanCovStoreFPointerFunction =
-      M.getOrInsertFunction(SanCovStoreFunPointer, VoidTy, PtrTy, IntptrTy);
+      M.getOrInsertFunction(SanCovStoreFunPointer, VoidTy, IntptrTy, IntptrTy);
 
   {
     AttributeList AL;
@@ -965,32 +965,6 @@ void ModuleSanitizerCoverage::InjectTraceForLoadsAndStores(
   }
 }
 
-/**
-/ Returns true only where the stored value is a pointer to a function.
-/ The stored value can be a statically known function, or it can come
-/ from somewhere like the field of a struct in allocated memory.
-/
-/ Looking at the Type of the stored Value by doing something like
-/  Store->getValueOperand()->stripPointerCasts()->getType()
-/  will fail to identify dynamic functions as a FunctionType because
-/  LLVM's FunctionType refers to statically known global functions,
-/  which the dynamic function by definition cannot be.
-/
-/ This is why we try to use debug information on the stored variable's
-/ supposed type instead.
-/
-/ TO DO: actually attempt the version that "doesn't work"
-**/
-static bool IsInterestingStore(StoreInst *Store) {
-
-  Value *StoreAddr = Store->getPointerOperand();
-
-  if () // To DO: use debug information
-    return true;
-  if () // To DO: use debug information
-    return true;
-  return false;
-}
 
 typedef std::vector<Type *> Types;
 typedef std::vector<int> Offsets;
@@ -1016,7 +990,7 @@ static retType RecursivelyHandleStructType(StructType *OriginalStruct) {
       if (this_types_offsets.size() == 0) {
         this_types_offsets.push_back(std::make_pair(Types{}, Offsets{}));
       }
-      for (int j = 0; j < this_types_offsets.size(); j++) {
+      for (unsigned j = 0; j < this_types_offsets.size(); j++) {
         Types this_types_j = this_types_offsets[j].first;
         Offsets this_offsets_j = this_types_offsets[j].second;
         this_types_j.insert(this_types_j.begin(), OriginalStruct);
@@ -1054,6 +1028,8 @@ static retType HandleTypeRecursive(Type *Ty) {
   else {
     // TODO: handle more types
     assert(false && "Unexpected type during function pointer analysis.\n");
+    // unreachable, but makes the compiler complain less
+    return(false, vectorTypesOffsets{});
   }
 }
 
@@ -1121,7 +1097,7 @@ getOffsetChainFromAllValues(std::vector<Value *> possibleStoredValues) {
   std::vector<retType> res = {};
   std::vector<int> positiveResultIdxs = {};
 
-  for (int i = 0; i < possibleStoredValues.size(); i++) {
+  for (unsigned i = 0; i < possibleStoredValues.size(); i++) {
     Type *StoredType = possibleStoredValues[i]->getType();
     retType partial_res = HandleTypeRecursive(StoredType);
 
@@ -1146,10 +1122,10 @@ getOffsetChainFromAllValues(std::vector<Value *> possibleStoredValues) {
 
 static Value *getFunctionPointerUsingOffsetChain(
     Value *V, Offsets offsetChain, Types typeChain,
-    InstrumentationIRBuilder Builder, DebugLoc loc);
+    IRBuilder<>  &Builder);
 
 static Value *gFPUOC_Struct(Value *V, Offsets offsetChain, Types typeChain,
-                            InstrumentationIRBuilder Builder) {
+                            IRBuilder<>  &Builder) {
   V = V->stripPointerCastsAndAliases();
 
   int offset = offsetChain[0];
@@ -1175,8 +1151,7 @@ static Value *gFPUOC_Struct(Value *V, Offsets offsetChain, Types typeChain,
 static Value *
 getFunctionPointerUsingOffsetChain(Value *V, Offsets offsetChain,
                                    Types typeChain,
-                                   InstrumentationIRBuilder Builder) {
-  Type *Ty = V->getType();
+                                   IRBuilder<>  &Builder) {
   if (offsetChain.size() == 0) {
     return V;
   } else {
@@ -1192,7 +1167,7 @@ void ModuleSanitizerCoverage::InjectTraceForFunctionPointerStores(
     Function &, ArrayRef<StoreInst *> Stores) {
 
   for (StoreInst *SI : Stores) {
-    InstrumentationIRBuilder IRB(SI);
+    IRBuilder<>  IRB(SI);
     Value *StorePtr = SI->getPointerOperand();
     Value *StoredValue = SI->getValueOperand();
 
@@ -1204,7 +1179,7 @@ void ModuleSanitizerCoverage::InjectTraceForFunctionPointerStores(
 
     // This means there is a function pointer somewhere
     if (offsetChainRes.first) {
-      for (int i = 0 : i < offsetChainRes.second.size(), i++) {
+      for (unsigned i = 0 ; i < offsetChainRes.second.size(); i++) {
         // (typeChain, offsetChain) are a chain of StructType and
         // indexes used to reach a function pointer from StoredValue
         Types typeChain = offsetChainRes.second[i].first;
-- 
2.43.0

