From 71b8cc4e251d79943d5480a6015e1979cd4d0b27 Mon Sep 17 00:00:00 2001
From: Daniel Wappner <dwappner@dc.uba.ar>
Date: Mon, 6 Apr 2026 09:46:27 -0300
Subject: [PATCH] some debug info and a bug fix

---
 .../Instrumentation/SanitizerCoverage.cpp     | 58 ++++++++++++++++++-
 1 file changed, 56 insertions(+), 2 deletions(-)

diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index 589ede67002c..bdae620f97a4 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -972,6 +972,38 @@ typedef std::vector<int> Offsets;
 typedef std::vector<std::pair<Types, Offsets>> vectorTypesOffsets;
 typedef std::pair<bool, vectorTypesOffsets> retType;
 
+static void showTypeAndOffsetChain(Types typeChain, Offsets offsetChain)
+{
+    errs() << "{\n";
+    for (unsigned j = 0; j < typeChain.size(); j++)
+    {
+        errs() << std::string('\t', j) << "Type: " << *typeChain[j] << " Offset: " << offsetChain[j] << "\n";
+    }
+    errs() << "}\n";
+}
+
+template <typename T>
+static void ShowVector(std::vector<T*> genericVector)
+{
+    errs() << "(size " << genericVector.size() << ") \n{\n";
+    for (unsigned i = 0; i < genericVector.size(); i++)
+    {
+        errs() << *genericVector[i] << "\n";
+    }
+    errs() << "}\n";
+}
+
+static void ShowTypesOffsets(vectorTypesOffsets typesOffsets)
+{
+    errs() << "(size " <<  typesOffsets.size() << ") \n{\n";
+    for (unsigned i = 0; i < typesOffsets.size(); i++)
+    {
+        showTypeAndOffsetChain(typesOffsets[i].first, typesOffsets[i].second);
+    }
+    errs() << "}\n";
+}
+
+
 static retType HandleTypeRecursive(Type *Ty);
 static retType RecursivelyHandlePointerType(PointerType *Ty) {
   Type *OrigType = Ty->getPointerElementType();
@@ -1102,6 +1134,12 @@ getOffsetChainFromAllValues(std::vector<Value *> possibleStoredValues) {
     Type *StoredType = possibleStoredValues[i]->getType();
     retType partial_res = HandleTypeRecursive(StoredType);
 
+    errs() << "*** VALUE " << *possibleStoredValues[i] << " OF TYPE " << *StoredType << " OBTAINED RESULT: " << partial_res.first << "\n";
+    errs() << "\t HAS POINTER: " << partial_res.first << "\n";
+    errs() << "\t OFFSET CHAINS: \n";
+    ShowTypesOffsets(partial_res.second);
+    errs() << "***\n";
+
     if (partial_res.first) {
       any += 1;
       positiveResultIdxs.push_back(i);
@@ -1112,12 +1150,11 @@ getOffsetChainFromAllValues(std::vector<Value *> possibleStoredValues) {
   // TO DO: not have this weakness
   assert(any <= 1 && "We cannot handle more than one intermeidate value "
                      "reaching a function pointer");
-  if (any > 1) {
+  if (any > 0) {
     // assume positiveResultIdxs.size() == 1.
     return std::make_pair(true, res[positiveResultIdxs[0]].second);
   } else {
     return std::make_pair(false, vectorTypesOffsets{});
-    ;
   }
 }
 
@@ -1167,7 +1204,17 @@ getFunctionPointerUsingOffsetChain(Value *V, Offsets offsetChain,
 void ModuleSanitizerCoverage::InjectTraceForFunctionPointerStores(
     Function &, ArrayRef<StoreInst *> Stores) {
 
+  errs() << "--- CALLED ON NEW BASIC BLOCK ---\n";
+
+
   for (StoreInst *SI : Stores) {
+    DebugLoc DL = SI->getDebugLoc();
+    std::string lineno;
+    if (DL.get())
+      lineno = std::to_string(DL.getLine());
+    else
+      lineno = "unknown";
+    errs() << "[LINE " << lineno << "] CURRENTLY ON INSTRUCTION " << *SI << "\n";
     IRBuilder<>  IRB(SI);
     Value *StorePtr = SI->getPointerOperand();
     Value *StoredValue = SI->getValueOperand();
@@ -1175,11 +1222,18 @@ void ModuleSanitizerCoverage::InjectTraceForFunctionPointerStores(
     // We consider all types this stored value had in the past.
     std::vector<Value *> StoredCastIntermediates =
         GetAllValuesThatWouldBeStripped(StoredValue);
+
+    errs() << "OBTAINED INTERMEDIATE VALUES: \n";
+    ShowVector(StoredCastIntermediates);
+
     retType offsetChainRes =
         getOffsetChainFromAllValues(StoredCastIntermediates);
 
     // This means there is a function pointer somewhere
     if (offsetChainRes.first) {
+      errs() << "INSTRUCTION " << *SI << " REACHES A FUNCTION POINTER. OFFSET CHAIN: ";
+      ShowTypesOffsets(offsetChainRes.second); errs() << "\n";
+
       for (unsigned i = 0 ; i < offsetChainRes.second.size(); i++) {
         // (typeChain, offsetChain) are a chain of StructType and
         // indexes used to reach a function pointer from StoredValue
-- 
2.43.0

