From 32ad6c09f9b80d901a9659b1d433cbfba2a3ea7f Mon Sep 17 00:00:00 2001
From: Daniel Wappner <dwappner@dc.uba.ar>
Date: Mon, 6 Apr 2026 14:15:13 -0300
Subject: [PATCH 2/3] gFPUOC_Struct: fix dangling dereference that made array
 pointer invalid

---
 .../Instrumentation/SanitizerCoverage.cpp     | 111 ++++++++----------
 1 file changed, 50 insertions(+), 61 deletions(-)

diff --git a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
index b5dd40a34ecf..e38451e916cd 100644
--- a/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
+++ b/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp
@@ -966,55 +966,46 @@ void ModuleSanitizerCoverage::InjectTraceForLoadsAndStores(
   }
 }
 
-
 typedef std::vector<Type *> Types;
 typedef std::vector<int> Offsets;
 typedef std::vector<std::pair<Types, Offsets>> vectorTypesOffsets;
 typedef std::pair<bool, vectorTypesOffsets> retType;
 
-static void showTypeAndOffsetChain(Types typeChain, Offsets offsetChain)
-{
-    errs() << "{\n";
-    for (unsigned j = 0; j < typeChain.size(); j++)
-    {
-        errs() << std::string('\t', j) << "Type: " << *typeChain[j] << " Offset: " << offsetChain[j] << "\n";
-    }
-    errs() << "}\n";
+static void showTypeAndOffsetChain(Types typeChain, Offsets offsetChain) {
+  errs() << "{\n";
+  for (unsigned j = 0; j < typeChain.size(); j++) {
+    errs() << std::string('\t', j) << "Type: " << *typeChain[j]
+           << " Offset: " << offsetChain[j] << "\n";
+  }
+  errs() << "}\n";
 }
 
 template <typename T>
-static void ShowPointerVector(std::vector<T*> genericVector)
-{
-    errs() << "(size " << genericVector.size() << ") \n{\n";
-    for (unsigned i = 0; i < genericVector.size(); i++)
-    {
-        errs() << *genericVector[i] << "\n";
-    }
-    errs() << "}\n";
+static void ShowPointerVector(std::vector<T *> genericVector) {
+  errs() << "(size " << genericVector.size() << ") \n{\n";
+  for (unsigned i = 0; i < genericVector.size(); i++) {
+    errs() << *genericVector[i] << "\n";
+  }
+  errs() << "}\n";
 }
 
 template <typename T>
-static void ShowElementVector(std::vector<T> genericVector)
-{
-    errs() << "(size " << genericVector.size() << ") \n{\n";
-    for (unsigned i = 0; i < genericVector.size(); i++)
-    {
-        errs() << genericVector[i] << "\n";
-    }
-    errs() << "}\n";
+static void ShowElementVector(std::vector<T> genericVector) {
+  errs() << "(size " << genericVector.size() << ") \n{\n";
+  for (unsigned i = 0; i < genericVector.size(); i++) {
+    errs() << genericVector[i] << "\n";
+  }
+  errs() << "}\n";
 }
 
-static void ShowTypesOffsets(vectorTypesOffsets typesOffsets)
-{
-    errs() << "(size " <<  typesOffsets.size() << ") \n{\n";
-    for (unsigned i = 0; i < typesOffsets.size(); i++)
-    {
-        showTypeAndOffsetChain(typesOffsets[i].first, typesOffsets[i].second);
-    }
-    errs() << "}\n";
+static void ShowTypesOffsets(vectorTypesOffsets typesOffsets) {
+  errs() << "(size " << typesOffsets.size() << ") \n{\n";
+  for (unsigned i = 0; i < typesOffsets.size(); i++) {
+    showTypeAndOffsetChain(typesOffsets[i].first, typesOffsets[i].second);
+  }
+  errs() << "}\n";
 }
 
-
 static retType HandleTypeRecursive(Type *Ty);
 static retType RecursivelyHandlePointerType(PointerType *Ty) {
   Type *OrigType = Ty->getPointerElementType();
@@ -1163,41 +1154,38 @@ getOffsetChainFromAllValues(std::vector<Value *> possibleStoredValues) {
   }
 }
 
-static Value *getFunctionPointerUsingOffsetChain(
-    Value *V, Offsets offsetChain, Types typeChain,
-    IRBuilder<>  &Builder);
+static Value *getFunctionPointerUsingOffsetChain(Value *V, Offsets offsetChain,
+                                                 Types typeChain,
+                                                 IRBuilder<> &Builder);
 
 static Value *gFPUOC_Struct(Value *V, Offsets offsetChain, Types typeChain,
-                            IRBuilder<>  &Builder) {
+                            IRBuilder<> &Builder) {
 
   errs() << "CALLED gFPUOC_STRUCT WITH VALUE BEFORE STRIP: " << *V << "\n";
 
   V = V->stripPointerCastsAndAliases();
 
   errs() << "VALUE AFTER STRIP: " << *V << "\n";
-  errs() << "OFFSET CHAIN: "; ShowElementVector(offsetChain);
-  errs() << " TYPE CHAIN: "; ShowPointerVector(typeChain);
+  errs() << "OFFSET CHAIN: ";
+  ShowElementVector(offsetChain);
+  errs() << " TYPE CHAIN: ";
+  ShowPointerVector(typeChain);
   errs() << "\n";
 
   int offset = offsetChain[0];
   Type *structType = typeChain[0];
 
-  ArrayRef<Value *> arrayParam = {
-      ConstantInt::get(Type::getInt32Ty(V->getContext()), 0),
-      ConstantInt::get(Type::getInt32Ty(V->getContext()), offset)};
-
-  
   unsigned AddrSpace = V->getType()->getPointerAddressSpace();
-  
+
   PointerType *StructPtrTy = PointerType::get(structType, AddrSpace);
   Value *Cast = Builder.CreatePointerCast(V, StructPtrTy);
   errs() << "GEP WITH STRUCT TYPE " << *structType << " AND offset " << offset
          << " ON VALUE " << *Cast << "\n";
-  errs() << "ARRAY PARAMS: (size " << arrayParam.size() << ")[";
-  for (Value *Param : arrayParam)
-    errs() << *Param << ", ";
-  errs() << "]\n";
-  Value *GEP = Builder.CreateInBoundsGEP(structType, Cast, arrayParam);
+
+  Value *GEP = Builder.CreateInBoundsGEP(
+      structType, Cast,
+      {ConstantInt::get(Type::getInt32Ty(V->getContext()), 0),
+       ConstantInt::get(Type::getInt32Ty(V->getContext()), offset)});
 
   Offsets remaining_offsets =
       Offsets(offsetChain.begin() + 1, offsetChain.end());
@@ -1206,10 +1194,9 @@ static Value *gFPUOC_Struct(Value *V, Offsets offsetChain, Types typeChain,
                                             remaining_types, Builder);
 }
 
-static Value *
-getFunctionPointerUsingOffsetChain(Value *V, Offsets offsetChain,
-                                   Types typeChain,
-                                   IRBuilder<>  &Builder) {
+static Value *getFunctionPointerUsingOffsetChain(Value *V, Offsets offsetChain,
+                                                 Types typeChain,
+                                                 IRBuilder<> &Builder) {
 
   if (offsetChain.size() == 0) {
     return V;
@@ -1227,7 +1214,6 @@ void ModuleSanitizerCoverage::InjectTraceForFunctionPointerStores(
 
   errs() << "--- CALLED ON NEW BASIC BLOCK ---\n";
 
-
   for (StoreInst *SI : Stores) {
     DebugLoc DL = SI->getDebugLoc();
     std::string lineno;
@@ -1235,8 +1221,9 @@ void ModuleSanitizerCoverage::InjectTraceForFunctionPointerStores(
       lineno = std::to_string(DL.getLine());
     else
       lineno = "unknown";
-    errs() << "[LINE " << lineno << "] CURRENTLY ON INSTRUCTION " << *SI << "\n";
-    IRBuilder<>  IRB(SI);
+    errs() << "[LINE " << lineno << "] CURRENTLY ON INSTRUCTION " << *SI
+           << "\n";
+    IRBuilder<> IRB(SI);
     Value *StorePtr = SI->getPointerOperand();
     Value *StoredValue = SI->getValueOperand();
 
@@ -1249,10 +1236,12 @@ void ModuleSanitizerCoverage::InjectTraceForFunctionPointerStores(
 
     // This means there is a function pointer somewhere
     if (offsetChainRes.first) {
-      errs() << "INSTRUCTION " << *SI << " REACHES A FUNCTION POINTER. OFFSET CHAIN: ";
-      ShowTypesOffsets(offsetChainRes.second); errs() << "\n";
+      errs() << "INSTRUCTION " << *SI
+             << " REACHES A FUNCTION POINTER. OFFSET CHAIN: ";
+      ShowTypesOffsets(offsetChainRes.second);
+      errs() << "\n";
 
-      for (unsigned i = 0 ; i < offsetChainRes.second.size(); i++) {
+      for (unsigned i = 0; i < offsetChainRes.second.size(); i++) {
         // (typeChain, offsetChain) are a chain of StructType and
         // indexes used to reach a function pointer from StoredValue
         Types typeChain = offsetChainRes.second[i].first;
-- 
2.43.0

